File: examples/mathematics/perturbation_strategies.rs
//! Compares different perturbation strategies for numerical Jacobian computation.

#[allow(unused_imports)]
use brahe::math::jacobian::{DNumericalJacobian, DJacobianProvider};
use nalgebra::{DMatrix, DVector};

fn main() {
    // Define dynamics with mixed-scale state
    let dynamics = |_t: f64, state: DVector<f64>| -> DVector<f64> {
        let x = state[0];
        let v = state[1];
        DVector::from_vec(vec![v, -x * 1e-6])
    };

    // Analytical Jacobian
    let analytical_jacobian = || -> DMatrix<f64> {
        DMatrix::from_row_slice(2, 2, &[
            0.0, 1.0,
            -1e-6, 0.0
        ])
    };

    // Test state with very different magnitudes
    let state = DVector::from_vec(vec![7000.0, 7.5]); // Position in km, velocity in km/s
    let t = 0.0;

    let j_analytical = analytical_jacobian();

    println!("Testing perturbation strategies on mixed-scale state:");
    println!("State: position={} km, velocity={} km/s\n", state[0], state[1]);

    // Strategy 1: Fixed perturbation
    println!("1. Fixed Perturbation (h = 1e-6)");
    let jacobian_fixed = DNumericalJacobian::central(Box::new(dynamics))
        .with_fixed_offset(1e-6);
    let j_fixed = jacobian_fixed.compute(t, state.clone());
    let error_fixed = (j_fixed - j_analytical.clone()).norm();
    println!("   Error: {:.2e}\n", error_fixed);

    // Strategy 2: Percentage perturbation
    println!("2. Percentage Perturbation (0.001%)");
    let jacobian_pct = DNumericalJacobian::central(Box::new(dynamics))
        .with_percentage(1e-5);
    let j_pct = jacobian_pct.compute(t, state.clone());
    let error_pct = (j_pct - j_analytical.clone()).norm();
    println!("   Error: {:.2e}\n", error_pct);

    // Strategy 3: Adaptive perturbation (recommended)
    println!("3. Adaptive Perturbation (scale=1.0, min=1.0)");
    let jacobian_adaptive = DNumericalJacobian::central(Box::new(dynamics))
        .with_adaptive(1.0, 1.0);
    let j_adaptive = jacobian_adaptive.compute(t, state.clone());
    let error_adaptive = (j_adaptive - j_analytical.clone()).norm();
    println!("   Error: {:.2e}\n", error_adaptive);

    // Summary
    println!("Strategy Comparison:");
    println!("  Fixed:      {:.2e}", error_fixed);
    println!("  Percentage: {:.2e}", error_pct);
    println!("  Adaptive:   {:.2e}", error_adaptive);
    println!("\nBest strategy: Adaptive (handles mixed scales robustly)");

    // Test with state component near zero
    println!("\n{}", "=".repeat(60));
    println!("Testing with component near zero:");
    let state_zero = DVector::from_vec(vec![7000.0, 1e-9]); // Very small velocity
    println!("State: position={} km, velocity={} km/s\n", state_zero[0], state_zero[1]);

    let j_analytical_zero = analytical_jacobian();

    // Percentage can be problematic when component is near zero
    let j_pct_zero = jacobian_pct.compute(t, state_zero.clone());
    let error_pct_zero = (j_pct_zero - j_analytical_zero.clone()).norm();
    println!("Percentage: Error = {:.2e}", error_pct_zero);

    // Adaptive handles it gracefully
    let j_adaptive_zero = jacobian_adaptive.compute(t, state_zero);
    let error_adaptive_zero = (j_adaptive_zero - j_analytical_zero).norm();
    println!("Adaptive:   Error = {:.2e}", error_adaptive_zero);

    println!("\nConclusion: Adaptive perturbation is most robust");
}

File: examples/mathematics/perturbation_strategies.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Compares different perturbation strategies for numerical Jacobian computation.

This example demonstrates how to choose appropriate step sizes for states
with different magnitudes.
"""

import brahe as bh
import numpy as np


# Define dynamics with mixed-scale state
# State: [large_position (km), small_velocity (km/s)]
def dynamics(t, state):
    # Simple dynamics with different scales
    x, v = state
    return np.array([v, -x * 1e-6])  # Different scales


# Analytical Jacobian
def analytical_jacobian(t, state):
    return np.array([[0.0, 1.0], [-1e-6, 0.0]])


# Test state with very different magnitudes
state = np.array([7000.0, 7.5])  # Position in km, velocity in km/s
t = 0.0

J_analytical = analytical_jacobian(t, state)

print("Testing perturbation strategies on mixed-scale state:")
print(f"State: position={state[0]} km, velocity={state[1]} km/s\n")

# Strategy 1: Fixed perturbation
print("1. Fixed Perturbation (h = 1e-6)")
jacobian_fixed = bh.NumericalJacobian.central(dynamics).with_fixed_offset(1e-6)
J_fixed = jacobian_fixed.compute(t, state)
error_fixed = np.linalg.norm(J_fixed - J_analytical)
print(f"   Error: {error_fixed:.2e}\n")

# Strategy 2: Percentage perturbation
print("2. Percentage Perturbation (0.001%)")
jacobian_pct = bh.NumericalJacobian.central(dynamics).with_percentage(1e-5)
J_pct = jacobian_pct.compute(t, state)
error_pct = np.linalg.norm(J_pct - J_analytical)
print(f"   Error: {error_pct:.2e}\n")

# Strategy 3: Adaptive perturbation (recommended)
print("3. Adaptive Perturbation (scale=1.0, min=1.0)")
jacobian_adaptive = bh.NumericalJacobian.central(dynamics).with_adaptive(
    scale_factor=1.0, min_value=1.0
)
J_adaptive = jacobian_adaptive.compute(t, state)
error_adaptive = np.linalg.norm(J_adaptive - J_analytical)
print(f"   Error: {error_adaptive:.2e}\n")

# Summary
print("Strategy Comparison:")
print(f"  Fixed:      {error_fixed:.2e}")
print(f"  Percentage: {error_pct:.2e}")
print(f"  Adaptive:   {error_adaptive:.2e}")
print("\nBest strategy: Adaptive (handles mixed scales robustly)")

# Test with state component near zero
print("\n" + "=" * 60)
print("Testing with component near zero:")
state_zero = np.array([7000.0, 1e-9])  # Very small velocity
print(f"State: position={state_zero[0]} km, velocity={state_zero[1]} km/s\n")

J_analytical_zero = analytical_jacobian(t, state_zero)

# Percentage fails when component is near zero
try:
    J_pct_zero = jacobian_pct.compute(t, state_zero)
    error_pct_zero = np.linalg.norm(J_pct_zero - J_analytical_zero)
    print(f"Percentage: Error = {error_pct_zero:.2e}")
except ZeroDivisionError:
    print("Percentage: FAILED (division by near-zero)")

# Adaptive handles it gracefully
J_adaptive_zero = jacobian_adaptive.compute(t, state_zero)
error_adaptive_zero = np.linalg.norm(J_adaptive_zero - J_analytical_zero)
print(f"Adaptive:   Error = {error_adaptive_zero:.2e}")

print("\nConclusion: Adaptive perturbation is most robust")
